### PR TITLE
Implementation of Weighted CRF Tagger (handling unbalanced datasets)

### LINKED ISSUE #4619: Handling unbalanced datasets in the CRF tagger
I'm using an unbalanced Corpus NER and I would like to add weights to the entities in the training step via `nn.CrossEntropyLoss`, I would like to know which .py file can I call my own lib and pass the weights to the model ???

**Some tutorials point to**: https://github.com/allenai/allennlp/blob/052353ed62e3a54fd7b39a660e65fc5dd2f91c7d/allennlp/nn/util.py#L628
**Version Allennlp**: 0.9.0
**Script**:

```
{
  "dataset_reader": {
    "type": "conll2003",
    "tag_label": "ner",
    "coding_scheme": "BIOUL",
    "token_indexers": {
      "tokens": {
        "type": "single_id",
        "lowercase_tokens": true
      },
      "token_characters": {
        "type": "characters",
        "min_padding_length": 3
      },
      "elmo": {
        "type": "elmo_characters"
     }
    }
  },
  "train_data_path": "train.txt",
  "validation_data_path": "dev.txt",
  "test_data_path": "test.txt",
  "model": {
    "type": "crf_tagger",
    "label_encoding": "BIOUL",
    "calculate_span_f1": true,
    "dropout": 0.5,
    "include_start_end_transitions": false,
    "text_field_embedder": {
      "token_embedders": {
        "tokens": {
            "type": "embedding",
            "embedding_dim": 300,
            "pretrained_file": "/model/glove/glove_s300.zip",
            "trainable": true
        },
        "elmo":{
          "type": "elmo_token_embedder",
          "options_file": "/model/elmo/options.json",
          "weight_file": "/model/elmo/elmo_pt_weights_dgx1.hdf5",
          "do_layer_norm": false,
          "dropout": 0.0
        },
        "token_characters": {
            "type": "character_encoding",
            "embedding": {
            "embedding_dim": 16
            },
            "encoder": {
            "type": "cnn",
            "embedding_dim": 16,
            "num_filters": 128,
            "ngram_filter_sizes": [3],
            "conv_layer_activation": "relu"
            }
        }
      }
    },
    "encoder": {
      "type": "lstm",
      "input_size": 1452,
      "hidden_size": 200,
      "num_layers": 2,
      "dropout": 0.5,
      "bidirectional": true
    },
    "verbose_metrics": true,
    "regularizer": [
      [
        "scalar_parameters",
        {
          "type": "l2",
          "alpha": 0.1
        }
      ]
    ]
  },
  "iterator": {
    "type": "basic",
    "batch_size":32
  },
  "trainer": {
    "optimizer": {
        "type": "adam",
        "lr": 0.001
    },
    "validation_metric": "+f1-measure-overall",
    "num_serialized_models_to_keep": 3,
    "num_epochs": 10,
    "grad_norm": 5.0,
    "patience": 25,
    "cuda_device":[0] 
  },
}
```

### PR DESCRIPTION
**Pull request by eraldoluis**
Closes #4619 .

Dependency of [allennlp-models PR #341](https://github.com/allenai/allennlp-models/pull/341)

Changes proposed in this pull request:

- I implemented and [experimentally compared](https://eraldoluis.github.io/2022/05/10/weighted-crf.html) three sample weighting strategies for CrfTagger.
- The three strategies are implemented in the files `allennlp/modules/conditional_random_field_**<strategy>**.py`, where **<strategy>** can be: `wemission`, `wtrans`, or `lannoy`.
- I generalized the methods `_input_likelihood(...)` and `_joint_likelihood(...)` of the `ConditionalRandomField` class so that they now receive an argument with the transition weights. In that way, I could implement the two basic sample weighting strategies (wemission and wtrans) by just subclassing this class and weighting the corresponding weights (`logits` and `transitions`) in the `forward(...)` method before calling `_input_likelihood(...)` and `_joint_likelihood(...)`. No modification was necessary to the basic algorithms in these two methods.
- On the other hand, the strategy proposed by [Lannoy et al.](https://perso.uclouvain.be/michel.verleysen/papers/ieeetbe12gdl.pdf) (suggestion by @dirkgr) needed a quite different implementation.

## Before submitting

- [x] I've read and followed all steps in the [Making a pull request](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/allenai/allennlp/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

## After submitting

- [ ] All GitHub Actions jobs for my pull request have passed.
- [ ] **`codecov/patch`** reports high test coverage (at least 90%).
    You can find this under the "Actions" tab of the pull request once the other checks have finished.

### CODE REVIEW

**Comment by epwalsh** [comment]
Hi @eraldoluis, thanks for this! I may not have time for a thorough review this week but this will be a priority next week.

**Comment by eraldoluis** [comment]
> Hi @eraldoluis, thanks for this! I may not have time for a thorough review this week but this will be a priority next week.

Thank you, @epwalsh !

File: allennlp/modules/conditional_random_field_wemission.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        label_weights: List[float],
+        constraints: List[Tuple[int, int]] = None,
+        include_start_end_transitions: bool = True,
+    ) -> None:
+        super().__init__(num_tags, constraints, include_start_end_transitions)
+
+        if label_weights is None:
+            raise ConfigurationError("label_weights must be given")
+
+        self.label_weights = torch.nn.Parameter(torch.Tensor(label_weights), requires_grad=False)
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_wemission.py:52]: I think it might be better to use [`self.register_buffer()`](https://pytorch.org/docs/stable/generated/torch.nn.Module.html#torch.nn.Module.register_buffer) here instead of defining the weights as a parameter. That way we can be sure the label weights aren't passed to the optimizer.

https://discuss.pytorch.org/t/what-is-the-difference-between-register-buffer-and-register-parameter-of-nn-module/32723/11

File: allennlp/modules/conditional_random_field_wemission.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+from .conditional_random_field import ConditionalRandomField
+
+
+class ConditionalRandomFieldWeightEmission(ConditionalRandomField):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_wemission.py:18]: Could you add something about the weighting strategy here?

File: allennlp/modules/conditional_random_field_wtrans.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+from .conditional_random_field import ConditionalRandomField
+
+
+class ConditionalRandomFieldWeightTrans(ConditionalRandomField):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_wtrans.py:18]: Should also have a note about the weighting strategy here.

File: allennlp/modules/conditional_random_field_lannoy.py [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        raise ConfigurationError(f"Unknown constraint type: {constraint_type}")
+
+
+class ConditionalRandomFieldLannoy(torch.nn.Module):
+    """
+    This module uses the "forward-backward" algorithm to compute
+    the log-likelihood of its inputs assuming a conditional random field model.
+
+    See, e.g. http://www.cs.columbia.edu/~mcollins/fb.pdf
+
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_lannoy.py:165]: Also would be good to have a note about the weight strategy here.

File: allennlp/modules/conditional_random_field_lannoy.py [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+import torch
+
+from allennlp.common.checks import ConfigurationError
+import allennlp.nn.util as util
+
+VITERBI_DECODING = Tuple[List[int], float]  # a list of tags, and a viterbi score
+
+
+def allowed_transitions(constraint_type: str, labels: Dict[int, str]) -> List[Tuple[int, int]]:
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_lannoy.py:14]: Is this any different from the same function in `conditional_random_field.py`?

File: allennlp/modules/conditional_random_field_lannoy.py [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+                to_entity = ""
+            else:
+                to_tag = to_label[0]
+                to_entity = to_label[1:]
+            if is_transition_allowed(constraint_type, from_tag, from_entity, to_tag, to_entity):
+                allowed.append((from_label_index, to_label_index))
+    return allowed
+
+
+def is_transition_allowed(
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_lannoy.py:59]: Same question here: is any different from the original?

File: allennlp/modules/conditional_random_field_lannoy.py [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+        # Include special start of sequence (num_tags + 1) and end of sequence tags (num_tags + 2)
+        if constraints is None:
+            # All transitions are valid.
+            constraint_mask = torch.full((num_tags + 2, num_tags + 2), 1.0)
+        else:
+            constraint_mask = torch.full((num_tags + 2, num_tags + 2), 0.0)
+            for i, j in constraints:
+                constraint_mask[i, j] = 1.0
+
+        self._constraint_mask = torch.nn.Parameter(constraint_mask, requires_grad=False)
```
Comment (epwalsh) [inline allennlp/modules/conditional_random_field_lannoy.py:210]: This should probably be a buffer as well, but I just realized this is how it's done in the original CRF module, so I guess I'm okay with this.

**Review by epwalsh — COMMENTED** [review]
Hey @eraldoluis, this looks really great! This is my first pass through (I still need to go through some of the code in more detail to get my head around it), but I just have some minor comments. In addition to the comments below, I will also say I think we should organize these modules into a common parent module. That is, create a new folder `allennlp/modules/conditional_random_field/` and move the 3 implementations into there.

File: allennlp/modules/conditional_random_field_wemission.py (hunk 1 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_wemission.py]: Done!

File: allennlp/modules/conditional_random_field_wemission.py (hunk 2 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_wemission.py]: Done!

File: allennlp/modules/conditional_random_field_wtrans.py (hunk 3 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_wtrans.py]: Done!

File: allennlp/modules/conditional_random_field_lannoy.py (hunk 4 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_lannoy.py]: Done!

File: allennlp/modules/conditional_random_field_lannoy.py (hunk 5 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_lannoy.py]: I simply forgot to adapt this class. I think now it is much better.

File: allennlp/modules/conditional_random_field_lannoy.py (hunk 6 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_lannoy.py]: Idem.

File: allennlp/modules/conditional_random_field_lannoy.py (hunk 7 above)
Comment (eraldoluis) [reply allennlp/modules/conditional_random_field_lannoy.py]: Ok. I didn't touch it. Although I agree with you that this should be a buffer as well.

**Comment by eraldoluis** [comment]
Thank you a lot @epwalsh for the effort you put on this.

I tried to address your first concerns. Let me know what you think about my changes.

I am looking forward to your feedback regarding the whole thing. Let me know if you have any questions. I will be happy to discuss this further if necassary.

**Comment by epwalsh** [comment]
Thanks for the quick responses/fixes! Changes look good. I should clarify what I meant by:

> create a new folder allennlp/modules/conditional_random_field/ and move the 3 implementations into there.

Looks like you left `allennlp/modules/conditional_random_field.py` where it is, and then moved the weighted CRFs into `allennlp/modules/conditional_random_field_weighted/`. I'd rather have a single submodule (folder) called `allennlp/modules/conditional_random_field/` with all of the CRFs (included the non-weighted base class).

**Comment by epwalsh** [comment]
I liked your blog post a lot by the way!

**Comment by eraldoluis** [comment]
> Looks like you left `allennlp/modules/conditional_random_field.py` where it is, and then moved the weighted CRFs into `allennlp/modules/conditional_random_field_weighted/`. I'd rather have a single submodule (folder) called `allennlp/modules/conditional_random_field/` with all of the CRFs (included the non-weighted base class).

Yes. I was unsure at first. But now I renamed the module to `conditional_random_field` and moved the original class to it. I also updated the changelog, which I had forgotten.

I also updated `allennlp-models` to reflect the new module organization. Unfortunately, I pushed first to the allennlp repository and the Model Tests failed (because `allennlp-models` was outdated). But these tests should pass now.

Let me know what do you think.

**Comment by eraldoluis** [comment]
Thank you very much, @epwalsh and @dirkgr ! This was my first contribution for an open source project and it was quite fun. I will definitely try it again soon. :)
